lib/pdfkit/source.rb

```diff
@@ -42,11 +42,45 @@ def to_s
     private
 
     def shell_safe_url
-      url_needs_escaping? ? URI::DEFAULT_PARSER.escape(@source) : @source
+      url = url_needs_escaping? ? URI::DEFAULT_PARSER.escape(@source) : @source
+
+      URI::DEFAULT_PARSER.parse(url)
+      shellescape_query(url)
     end
 
     def url_needs_escaping?
       URI::DEFAULT_PARSER.escape(URI::DEFAULT_PARSER.unescape(@source)) != @source
     end
+
+    def needs_shell_escaping?(data)
+      Shellwords.escape(shellwords_unescape(data)) != data
+    end
+
+    def shellescape_query(url)
+      url, query = url.split('?')
```

Comment:
You're assuming a properly formatted URL. That's a dangerous thing...What's the result when someone passes in url = `http://localhost:3000/haxor$HOME/hello`. Hint: It doesn't escape the important part....